### PR TITLE
feat: add skeleton animation & vertex aimation

### PR DESCRIPTION
**Pull request by JiepengTan**
project file 
[06-Skeleton.zip](https://github.com/user-attachments/files/16759760/06-Skeleton.zip)

config file : "assets/sprites/Tom/index.json" 
skeleton animation
```
{

   ... // other fields
  "animator": "skeleton.animator",
  "avatar": "skeleton.avatar",
   ... // other fields
}
```
skeleton animation

vertex animation 
```
{
   ... // other fields
  "animator": "vertex.animator",
  "avatar": "vertex.avatar",
   ... // other fields
}
```

animated_drawing

### CODE REVIEW

File: internal/anim/common/config.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	Image  string         `json:"image"`
+	Mesh   string         `json:"mesh"`
+	Scale  math32.Vector2 `json:"scale"`
+	Offset math32.Vector2 `json:"offset"`
+}
+
+type AnimatorConfig struct {
+	Name        string           `json:"name"`
+	Type        string           `json:"type"`
+	DefaultClip string           `json:"defaultClip"`
```
[the remark this answers is not in the document]
Comment (nighca) [reply internal/anim/common/config.go:17]: And is there any alternative to `animBindings`?

File: internal/anim/anim.go [hunk 2]
```diff
@@ -72,6 +98,41 @@ type Anim struct {
...
+	}
+
+	var avatarConfig common.AvatarConfig
+	err = common.LoadJson(&avatarConfig, fs, spriteDir, avatarPath)
+	if err != nil {
+		log.Panicf("avatar config [%s] not exist", animatorPath)
+	}
+
+	var animator IAnimator
+	if config.Type == ANIMATOR_TYPE_VERTEX {
```
[the remark this answers is not in the document]
Comment (JiepengTan) [reply internal/anim/anim.go:115]: for now, we will not expose this concept to the users, they will not be aware of this implementation detail.

File: internal/anim/common/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JiepengTan) [reply internal/anim/common/config.go:18]: Yes, the priority of sprite.defaultAnimation is higher than animator.defaultClip. Currently, this animation system is still in the experimental stage and is not intended to be made available to users. This PR is to support the development of Builder, which depends on skeletal animations.

[the remark this answers is not in the document]
Comment (nighca) [reply internal/anim/common/config.go:18]: > Currently, this animation system is still in the experimental stage and is not intended to be made available to users. This PR is to support the development of Builder, which depends on skeletal animations.

Sure.

As there's no sample projects in this PR now, I'm not sure:

1. Is it supported for one sprite to contain both some frame-animations and some skeletal-animations?
2. How are we supposed do animation binding with skeletal-animations? For example, bind clip "dying" to behavior `die`? Do we use `animBindings: { die: "dying" }` in sprite config just like frame-animations?

[the remark this answers is not in the document]
Comment (JiepengTan) [reply internal/anim/common/config.go:18]: project file 
[06-Skeleton.zip](https://github.com/user-attachments/files/16759760/06-Skeleton.zip)

File: internal/anim/anim.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nighca) [reply internal/anim/anim.go:120]: The users will notice that there's `animator` (with `type: skeletal | vertex`) in sprite config to support skeletal or vertex animations.

In the long term, is it in plan to move config of frame-animations to some "animator file", just like skeletal & vertex animation?

File: internal/anim/common/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JiepengTan) [reply internal/anim/common/config.go:18]: > > Currently, this animation system is still in the experimental stage and is not intended to be made available to users. This PR is to support the development of Builder, which depends on skeletal animations.
> 
> Sure.
> 
> As there's no sample projects in this PR now, I'm not sure:
> 
> 1. Is it supported for one sprite to contain both some frame-animations and some skeletal-animations?
> 2. How are we supposed do animation binding with skeletal-animations? For example, bind clip "dying" to behavior `die`? Do we use `animBindings: { die: "dying" }` in sprite config just like frame-animations?

@nighca 
The skeletal animation at the current stage is not intended to be open to users, but is designed to facilitate the development of the **"Training camp"**. If it becomes available in the future, we will add a frame Animator to ensure that skeleton animation is consistent with frame animation.

[the remark this answers is not in the document]
Comment (nighca) [reply internal/anim/common/config.go:18]: As discussed in the voice conference, this PR is not intended for merging into the `main` branch, as its APIs are not yet stable.

Similar to the `trailhead_asset` branch of the builder, the "Training Camp" development will maintain dependencies on a standalone branch of spx.

**Pull request closed without merging**